gqcp/src/ONVBasis/SeniorityZeroONVBasis.cpp

```diff
@@ -55,179 +57,323 @@ size_t SeniorityZeroONVBasis::calculateDimension(const size_t K, const size_t N_
+        }
+    }  // address (I) loop
+}
+
+
+/*
+ *  MARK: Dense restricted operator evaluations
+ */
+
+/**
+ *  Calculate the dense matrix representation of a Hubbard Hamiltonian in this ONV basis.
  *
- *  @param one_op               the one-electron operator in an orthonormal orbital basis to be evaluated in the ONV basis
+ *  @param hamiltonian      A Hubbard Hamiltonian expressed in an orthonormal orbital basis.
  *
- *  @return the operator's diagonal evaluation in a vector with the dimension of the ONV basis
+ *  @return A dense matrix represention of the Hamiltonian.
+ */
+SquareMatrix<double> SeniorityZeroONVBasis::evaluateOperatorDense(const RSQHamiltonian<double>& hamiltonian) const {
+
+    if (hamiltonian.numberOfOrbitals() != this->numberOfSpatialOrbitals()) {
+        throw std::invalid_argument("SeniorityZeroONVBasis::evaluateOperatorDense(const RSQHamiltonian<double>&): The number of spatial orbitals for the ONV basis and Hamiltonian are incompatible.");
+    }
+
+
+    // Prepare some variables to be used in the algorithm.
+    const size_t N_P = this->numberOfElectronPairs();
+    const size_t dim = this->dimension();
+
+    SquareMatrix<double> H = SquareMatrix<double>::Zero(dim);  // The matrix representation of the Hamiltonian.
+    const auto diagonal = this->evaluateOperatorDiagonal(hamiltonian);
+    const auto& g = hamiltonian.twoElectron().parameters();
+
+
+    // Use a proxy ONV basis to treat alpha- and beta- ONVs as equal and multiply all contributions by 2.
+    const auto proxy_onv_basis = this->proxy();
+    auto onv = proxy_onv_basis.constructONVFromAddress(0);  // Create the ONV with address 0.
+    for (size_t I = 0; I < dim; I++) {                      // I loops over all the addresses of the ONVs
+
+        H(I, I) += diagonal(I);
+
+        for (size_t e1 = 0; e1 < N_P; e1++) {            // e1 (electron 1) loops over the number of electrons
+            const size_t p = onv.occupationIndexOf(e1);  // retrieve the index of the orbital that the electron occupies
+
+            // Remove the weight from the initial address I, because we annihilate
+            size_t address = I - proxy_onv_basis.vertexWeight(p, e1 + 1);
+
+            // The e2 iteration counts the number of encountered electrons for the creation operator.
+            // We only consider greater addresses than the initial one (because of symmetry), hence we only count electron after the annihilated electron (e1).
+            size_t e2 = e1 + 1;
+            size_t q = p + 1;
+
+            // perform a shift  TODO: clarify what shift
+            proxy_onv_basis.shiftUntilNextUnoccupiedOrbital<1>(onv, address, q, e2);
+
+            while (q < K) {
+                const size_t J = address + proxy_onv_basis.vertexWeight(q, e2);
+
+                H(I, J) += g(p, q, p, q);
+                H(J, I) += g(p, q, p, q);
+
+                q++;  // go to the next orbital
+
+                // perform a shift  TODO: clarify what shift
```

Comment:
```suggestion
                // Perform a shift;  TODO: clarify what shift.
```

Same question as before. If it's on purpose, ignore this 😄